content/blog/2020/08/2020-08-29-git-performance-improvement-phase3.adoc

```diff
@@ -0,0 +1,126 @@
+
+* The performance of git fetch (average execution time/op) is strongly correlated to the size of a repository
+* There exists an inflection point on the scale of repository size after which the nature of `JGit` performance changes (it starts to degrade)
+* After running multiple benchmarks, it is safe to say that for a large sized repository `command line git` would be a better choice of implementation.
+* We can use this insight to implement a feature which avoids `JGit` when it comes to large repositories.
+
+=== Stage 2: Validate the benchmark results across multiple platforms
+
+Running git fetch operation for a 400 MiB sized repository on:
+
+* `Windows`
+* `FreeBSD-12`
+* `ppc64le`
+* `s390x`
+
+The result of running this experiment is given below:
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-multiple-platforms.png[Performance on multiple platforms]
+
+
+*Observation*: The difference in performance between `git` and `JGit` remains constant across all platforms which means any insight taken from the benchmarks to develop a solution will be platform agnostic in general.
+
+=== Stage 3: Performance of git fetch and repository structure
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-repo-diagram.png[title="Relationship of repository structure with performance" role="center", width=420]
+
+The area of the circle enclosing each parameter signifies the strength of the positive correlation between the performance of a git fetch operation and that parameter. From the diagram: +
+
+* Size of the aggregated objects is the *dominant* player in determining the execution time for a git fetch
+* Number of branches and Number of tags play a similar role but are strongly overshadowed by size of repository
+* Number of commits has a negligible effect on the performance of running git fetch
+
+*After running these experiments from Stage-1 to Stage-3, we developed a solution called the `GitToolChooser` which is explained in the next stage*
+
+=== Stage 4: Enabling faster checkouts with an automatic git tool chooser
+
+This feature takes the responsibility of choosing the optimal implementation from the *user* and hands it to the *plugin*. It takes the decision of recommending an implementation on the basis of the size of the repository. link:/blog/2020/07/2020-07-29-git-performance-improvement-phase2.adoc[Here is how it works].
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-perf-improv.png[title="Faster checkouts with GitToolChooser" role="center", width=938]
+
+The image above depicts the performance enhancements we have performed over the course of the GSoC project. These improvements have enabled the checkout step to be finished *within half of what it used to take* earlier in some cases.
+
+Let's talk about performance improvements in two parts.
+
+== Major Performance Improvements
```

Comment:
The preceding headings are using "sentence case" (capitalize first word, rest lower case).  Make this sentence case as well.
```suggestion
== Major performance improvements
```